### PR TITLE
[Auto Parallel] Add paddle.distributed.to_static api

### PR DESCRIPTION
**Pull request by pkuzyc**
### PR types
New features

### PR changes
APIs

### Description
Pcard-76459
Add paddle.distributed.to_static api and its returned class DistModel for converting dygraph auto parallel model to static mode.

paddle.distributed.to_static converts the model and data loader used in dygraph auto-parallelism to that in static mode auto-parallelism. It returns a DistModel instance that provides APIs and a DistributedDataLoader to generate data for static mode auto-parallel training, evaluation and prediction.

Doc: https://github.com/PaddlePaddle/docs/pull/6357

### CODE REVIEW

File: python/paddle/distributed/auto_parallel/api.py [hunk 1]
```diff
@@ -90,9 +91,263 @@ def sharding_specs(self):
...
+        if len(feed_name_list) != len(data_list):
+            raise ValueError(
+                "The input data and feed_list are not consistent."
+                "The model takes %s as input" % (str(feed_name_list))
+            )
+        return dict(zip(feed_name_list, data_list))
+
+    def __call__(self, *args):
+        if self._mode is None:
+            raise ValueError("Please call train()/eval()/predict() first.")
```
Comment (zhiqiu) [inline python/paddle/distributed/auto_parallel/api.py:260]: default is train

File: python/paddle/distributed/auto_parallel/api.py [hunk 2]
```diff
@@ -90,9 +91,263 @@ def sharding_specs(self):
...
+    # get the static graph in predict mode
+    dist_model._engine.prepare(
+        inputs_spec, None, mode="predict", init_parameters=False
+    )
+
+    # get DistributedDataLoader for static mode auto-parallelism
+    batch_size = dist_model._engine._validate_batch_size(batch_size)
+    dist_loader = dist_model._engine._prepare_dataloader(
+        loader.dataset, return_list=True, batch_size=batch_size
+    )
```
Comment (zhiqiu) [inline python/paddle/distributed/auto_parallel/api.py]: Better move these lines into `__init__ of DistModel`

File: python/paddle/distributed/auto_parallel/static/engine.py [hunk 3]
```diff
@@ -1545,6 +1559,29 @@ def run(self, data=None, feed=None, fetch_list=None, mode=None):
...
+        dist_main_prog = dist_context.dist_main_programs[self._cur_rank]
+        dist_startup_prog = dist_context.dist_startup_programs[self._cur_rank]
+        dist_main_block = dist_main_prog.global_block()
+
+        # NOTE: Get feed_list, then insert dataloader op with sharded var shape.
+        # Cause predict_program does not contain labels var,
+        # then we will add labels var from serial_program to dist_program,
+        # that maintains the length of feed_list equal to the length of dataset's values.
+        inputs_var = dist_context.serial_feed_vars["inputs"]
+        labels_var = dist_context.serial_feed_vars["labels"]
```
Comment (zhiqiu) [inline python/paddle/distributed/auto_parallel/static/engine.py:1544]: what if the feed is not called `inputs` and `labels`?

File: python/paddle/distributed/auto_parallel/api.py (hunk 1 above)
Comment (pkuzyc) [reply python/paddle/distributed/auto_parallel/api.py:260]: Done, default mode is set according to the init args.

File: python/paddle/distributed/auto_parallel/api.py (hunk 2 above)
Comment (pkuzyc) [reply python/paddle/distributed/auto_parallel/api.py]: Done.

File: python/paddle/distributed/auto_parallel/static/engine.py (hunk 3 above)
Comment (pkuzyc) [reply python/paddle/distributed/auto_parallel/static/engine.py:1544]: ``inputs`` and  ``labels`` are the key names of the dict ``dist_context.serial_feed_vars``, not the names of model input and label.

**Review by zhiqiu — DISMISSED** [review]
LGTM

File: python/paddle/distributed/auto_parallel/static/engine.py [hunk 4]
```diff
@@ -812,13 +812,27 @@ def _init_comm(self):
     def _share_parameters(self):
         # mapping from {variable -> parameter}
         named_params = self.program_helper.named_parameters()
+        dist_context = self._dist_contexts[self._mode]
+        dist_main_program = dist_context.dist_main_programs[self._cur_rank]
+
         for name, param in named_params.items():
             var = global_scope().var(name)
+            dense_tensor = var.get_tensor()
```
Comment (JZ-LIANG) [inline python/paddle/distributed/auto_parallel/static/engine.py]: miss a filter that filter out not to share param not in this rank in PP.

Comment (pkuzyc) [reply python/paddle/distributed/auto_parallel/static/engine.py]: Now the parameter initialization is moved to LayerHelper.init(), the filter is included in LayerHelper.init()

File: python/paddle/distributed/auto_parallel/api.py [hunk 5]
```diff
@@ -90,9 +91,324 @@ def sharding_specs(self):
...
+            >>> loader = paddle.io.DataLoader(dataset, batch_size=BATCH_SIZE)
+
+            >>> mesh = dist.ProcessMesh([0, 1], dim_names=["x"])
+            >>> layer = DemoNet(mesh)
+            >>> opt = paddle.optimizer.SGD(
+            ...     learning_rate=0.1, parameters=layer.parameters()
+            ... )
+            >>> loss_fn = nn.MSELoss()
+
+            >>> dist_model, dist_loader = dist.static_decorate(
```
Comment (jeff41404) [inline python/paddle/distributed/auto_parallel/api.py]: `dist.static_decorate` should be `dist.to_static` ?

File: python/paddle/distributed/auto_parallel/api.py [hunk 6]
```diff
@@ -90,9 +91,324 @@ def sharding_specs(self):
...
+def to_static(
+    layer: paddle.nn.Layer,
+    loader=None,
+    loss=None,
+    optimizer=None,
+    strategy=None,
+):
+    """
+    Converts the model and data loader used in dygraph auto-parallelism to
+    that in static mode auto-parallelism. static_decorate returns a DistModel
```
Comment (jeff41404) [inline python/paddle/distributed/auto_parallel/api.py]: `static_decorate ` should be `to_static` ?

File: test/auto_parallel/semi_auto_parallel_dist_to_static_api.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+            dist_model.train()
+        with self.assertRaises(RuntimeError):
+            dist_model.eval()
+        with self.assertRaises(RuntimeError):
+            dist_model.predict()
+        dist_model._engine._has_prepared["train"] = True
+        dist_model._engine._has_prepared["eval"] = True
+        dist_model._engine._has_prepared["predict"] = True
+
+    # python -m paddle.distributed.launch --devices=0,1 semi_auto_parallel_static_decorate_api.py
```
Comment (jeff41404) [inline test/auto_parallel/semi_auto_parallel_dist_to_static_api.py]: `semi_auto_parallel_static_decorate_api.py` should be `semi_auto_parallel_dist_to_static_api.py` ?

File: python/paddle/distributed/auto_parallel/api.py (hunk 5 above)
Comment (pkuzyc) [reply python/paddle/distributed/auto_parallel/api.py]: Done.

File: python/paddle/distributed/auto_parallel/api.py (hunk 6 above)
Comment (pkuzyc) [reply python/paddle/distributed/auto_parallel/api.py]: Done

File: test/auto_parallel/semi_auto_parallel_dist_to_static_mlp.py [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+        )
+
+        dy_losses = self.run_dynamic(dy_layer, dy_opt, data_loader)
+        dy2static_losses = self.run_dy2static(
+            dy2static_layer, dy2static_opt, data_loader
+        )
+
+        np.testing.assert_allclose(dy_losses, dy2static_losses, rtol=1e-6)
+
+    # python -m paddle.distributed.launch --devices=0,1 semi_auto_parallel_static_decorate_mlp.py
```
Comment (jeff41404) [inline test/auto_parallel/semi_auto_parallel_dist_to_static_mlp.py]: `semi_auto_parallel_static_decorate_mlp.py` should be `semi_auto_parallel_dist_to_static_mlp.py`?

File: python/paddle/distributed/auto_parallel/api.py [hunk 9]
```diff
@@ -90,9 +91,324 @@ def sharding_specs(self):
...
 # Part2: DistTensor construction related APIs
 
 
+def to_static(
+    layer: paddle.nn.Layer,
+    loader=None,
+    loss=None,
+    optimizer=None,
+    strategy=None,
+):
```
Comment (jeff41404) [inline python/paddle/distributed/auto_parallel/api.py:286]: I saw in the design document that there is parameter of `metrics`. Shall we need to implement `metrics` which is not implemented here? If not, please explain the reason and modify the design document.

File: test/auto_parallel/semi_auto_parallel_dist_to_static_api.py (hunk 7 above)
Comment (pkuzyc) [reply test/auto_parallel/semi_auto_parallel_dist_to_static_api.py]: Done.

File: test/auto_parallel/semi_auto_parallel_dist_to_static_mlp.py (hunk 8 above)
Comment (pkuzyc) [reply test/auto_parallel/semi_auto_parallel_dist_to_static_mlp.py]: Done.

File: python/paddle/distributed/auto_parallel/api.py (hunk 9 above)
Comment (pkuzyc) [reply python/paddle/distributed/auto_parallel/api.py:286]: ``metrics`` is not consisted in the original design, it is not used now. So removed it here. I will modify the design document.

**Review by zhiqiu — APPROVED** [review]
LGTM

File: python/paddle/distributed/auto_parallel/api.py [hunk 10]
```diff
@@ -90,9 +91,324 @@ def sharding_specs(self):
         return self._sharding_specs
 
 
+class DistModel:
+    """
+    DistModel is a wrapper of the network model for the use of static mode
+    auto parallel. DistModel contains the distributed Graph of the model and
+    offers the APIs for training, evaluation and prediction.
```
Comment (jzhang533) [inline python/paddle/distributed/auto_parallel/api.py:98]: is it possible to make this docstring more understandable ? 
e.g.: very challenging for me to understand the term : "static mode auto parallel.", even after googling this phrase.

**Review by jeff41404 — APPROVED** [review]
LGTM

File: python/paddle/distributed/auto_parallel/api.py [hunk 11]
```diff
@@ -90,9 +91,324 @@ def sharding_specs(self):
...
+            return outs["outputs"]
+        else:
+            return outs["loss"]
+
+
 # Part2: DistTensor construction related APIs
 
 
+def to_static(
+    layer: paddle.nn.Layer,
```
Comment (jzhang533) [inline python/paddle/distributed/auto_parallel/api.py:281]: I'd like to suggest not using type annotation for any arguments.

**Review by XieYunshen — APPROVED** [review]
LGTM for `set_tests_properties(test_semi_auto_parallel_dist_to_static
                       PROPERTIES LABELS "RUN_TYPE=EXCLUSIVE" TIMEOUT 300)`

File: python/paddle/distributed/auto_parallel/api.py [hunk 12]
```diff
@@ -90,9 +91,324 @@ def sharding_specs(self):
...
+            ...     # where the value of "out0" is the first output.
+            ...     outs = dist_model(image)
+
+            >>> # This case need to be excuted in multi-card environment
+            >>> # export CUDA_VISIBLE_DEVICES=0,1
+            >>> # python -m paddle.distributed.launch {test_case}.py
+    """
+    dist_model = DistModel(layer, loader, loss, optimizer, strategy)
+    dist_loader = dist_model.dist_loader
+
```
Comment (jzhang533) [inline python/paddle/distributed/auto_parallel/api.py:408]: the API name  `to_static` made me very confused, especially trying very hard to understand the relation to `paddle.jit.to_static`.

from this 2 LOC implementation, I think this API basically is a creator API for `DistModel`. I'd like to suggest a more intuitive API name, e.g.: dist_model_creator.

Comment (pkuzyc) [reply python/paddle/distributed/auto_parallel/api.py:408]: The function (or purpose) of this api is to convert a model whose parameters are Distributed Tensor (generated by ``shard_tensor``). I think ``to_static`` is more suitable for this function.

File: python/paddle/distributed/auto_parallel/api.py (hunk 10 above)
Comment (pkuzyc) [reply python/paddle/distributed/auto_parallel/api.py:98]: I will rephrase the docstring, update in next PR.

File: python/paddle/distributed/auto_parallel/api.py (hunk 11 above)
Comment (pkuzyc) [reply python/paddle/distributed/auto_parallel/api.py:281]: Update in next PR

**Review by jzhang533 — APPROVED** [review]
giving an approve, since we want to land this PR before release/2.6 branch cutoff in offline discussion with @zhiqiu and @pkuzyc .

will leave docstr improvement as future work.
